[JIT] X64 - Extend emitter peephole optimization of eliminating unnecessary `mov` instructions

TIHan opened the pull request:
**Description**
-
Resolves https://github.com/dotnet/runtime/issues/10315 based on the example code given in the first post.

This will not eliminate all possible unnecessary `mov` instructions, but it handles more now.

The JIT's emitter already had an existing way of removing the instructions using:
```cpp
bool emitter::AreUpper32BitsZero(regNumber reg)
```
But, it was only able to look back at one instruction.

This PR extends `AreUpper32BitsZero` to allow looking back up to 256 instructions(max limit of instructions for an IG).

Example diffs:
```diff
@@ -28,13 +28,9 @@ G_M17551_IG02:        ; gcrefRegs=00000000 {}, byrefRegs=00000004 {rdx}, byref
        movzx    r8, byte  ptr [rcx+01H]
        movzx    r9, byte  ptr [rcx+02H]
        movzx    rcx, byte  ptr [rcx+03H]
-       mov      eax, eax
        movsx    rax, byte  ptr [rdx+rax]
-       mov      r8d, r8d
        movsx    r8, byte  ptr [rdx+r8]
-       mov      r9d, r9d
        movsx    r9, byte  ptr [rdx+r9]
-       mov      ecx, ecx
        movsx    rdx, byte  ptr [rdx+rcx]
        ; byrRegs -[rdx]
        shl      eax, 18
@@ -43,12 +39,12 @@ G_M17551_IG02:        ; gcrefRegs=00000000 {}, byrefRegs=00000004 {rdx}, byref
        or       eax, edx
        or       r8d, r9d
        or       eax, r8d
-						;; size=66 bbWeight=1    PerfScore 27.25
+						;; size=56 bbWeight=1    PerfScore 26.25
 G_M17551_IG03:        ; , epilog, nogc, extend
        ret      
 						;; size=1 bbWeight=1    PerfScore 1.00
 
-; Total bytes of code 67, prolog size 0, PerfScore 34.95, instruction count 19, allocated bytes for code 67 (MethodHash=ea74bb70) for method System.Buffers.Text.Base64:Decode(ulong,byref):int
+; Total bytes of code 57, prolog size 0, PerfScore 32.95, instruction count 15, allocated bytes for code 57 (MethodHash=ea74bb70) for method System.Buffers.Text.Base64:Decode(ulong,byref):int
```

Diffs from the issue's example:
![image](https://user-images.githubusercontent.com/1278959/206354847-20d4d43f-dab1-4070-9cba-d1de2d6d965c.png)

**Acceptance Criteria**
-
- [x] Disasm test case
- [x] Comments

TIHan committed:
Added regUpper32BitsZeroLookup

TIHan committed:
Minor refactoring

TIHan committed:
Refactoring

TIHan committed:
Refactoring

TIHan committed:
Refactoring

TIHan committed:
Added comments, more refactoring

TIHan committed:
Added some logs

TIHan committed:
Added some logs

TIHan committed:
Added some logs

TIHan committed:
Added some logs

TIHan committed:
Remove logs. Fixed an issue.

TIHan committed:
Expanding peephole opt even more. Trying to improve tp.

TIHan committed:
Some refactoring

TIHan committed:
Trying to make this safe

TIHan committed:
work

TIHan committed:
more work

TIHan committed:
Fixing a few issues

TIHan committed:
Fixing a bug

TIHan committed:
Switching to a list of instrs

TIHan committed:
Adding simple peephole API

TIHan committed:
minor change

TIHan committed:
Trying to improve tp

TIHan committed:
Trying to improve tp

TIHan committed:
Fixed a minor issue

TIHan committed:
Small correctness issue

TIHan committed:
Minor change

TIHan committed:
Using emitCurIGinsCnt

TIHan committed:
Added EMIT_MAX_IG_INS_COUNT

markples commented on src/coreclr/jit/emitxarch.cpp:
```diff
@@ -478,70 +478,162 @@ bool emitter::IsFlagsAlwaysModified(instrDesc* id)
-
 bool emitter::AreUpper32BitsZero(regNumber reg)
 {
+#ifdef TARGET_AMD64
+    if (!((reg >= REG_RAX) && (reg <= REG_XMM15)))
+        return false;
+#else
+    if (!((reg >= REG_EAX) && (reg <= REG_XMM7)))
+        return false;
+#endif // !TARGET_AMD64
```
minor - I think that you could remove this `ifdef` by using `REG_FP_LAST`

TIHan replied:
Yea, looks like I can use this, then I can use `REG_EAX` as it is just an alias for `REG_RAX`

TIHan committed:
Formatting

TIHan committed:
Feedback

TIHan committed:
Feedback

TIHan committed:
Feedback

TIHan committed:
Minor change

TIHan committed:
More comments, minor change

TIHan committed:
Get rid of these asserts

TIHan committed:
Improving TP even further

tannergooding commented on src/coreclr/jit/emitxarch.cpp:
```diff
@@ -478,70 +478,150 @@ bool emitter::IsFlagsAlwaysModified(instrDesc* id)
-        case IF_RWR_SRD:
-        case IF_RWR_ARD:
+    emitPeepholeLastInstrs([&](instrDesc* id) {
+        switch (id->idInsFmt())
```
If this is going to be depended on like this, we may need to consider auditing `instrsxarch.h` and `emitfmtsxarch.h`

There are some cases like `emitIns_R_R_I` where it's using `IF_RRW_RRW_CNS` but such an instruction doesn't exist. It should actually be either `IF_RWR_RRD_CNS` or `IF_RRW_RRD_CNS` depending on which instruction is being used.

Likewise, just glancing through `instrsxarch.h` there are many instructions which aren't correctly tracking `IUM`. For example, many of the `SIMD` instructions are listed as `IUM_WR` which while "technically correct" for the VEX encoding, will result in an incorrect format for the legacy encoding.

This is because many SIMD instructions are `RMW` in the legacy encoding but can encode an additional register for the VEX encoding.

So, for example, `addps` is encoded as `addps rwreg, rdreg/rdmem` with the legacy encoding, but as `addps wrreg, rdreg, rdreg/rdmem` for the VEX encoding.

For most SIMD instructions we go through `emitIns_SIMD_*` which will detect if we're doing the VEX encoding or not and select the right path to go down. For example, `emitIns_SIMD_R_R_R` will go to `emitIns_R_R_R` for `VEX` and emit an elideable `mov` + `emitIns_R_R` for the legacy path.

`emitIns_R_R_R` hardcodes the format to `IF_RWR_RRD_RRD` so it is already correct.
`emitIns_R_R` on the other hand uses `emitInsModeFormat(ins, IF_RRD_RRD)` which will then see `IUM_WR` which is incorrect as that will resolve `IF_RWR_RRD` when it should've resolved `IF_RRW_RRD`.

Just glancing through, it looks like most `binary` instructions are incorrect. As are the handful of `ternary` instructions.
Some `unary` instructions are correct, but others (like `movss`) don't account for the merge semantics that may exist.

There are also a handful of VEX encoded instructions (like `vfmaddsub132pd`) which are still RMW and which are `IF_RRW_RRD_RRD` (`ins rwreg, rdreg, rdreg/rdmem`) and that are also not tracking the `IUM` correctly

TIHan replied:
We probably do need to vet these, but since `AreUpper32BitsZero` is only going to operate on GPRs, we don't have to worry too much about SIMD.

Also, I did add a few extra checks for instructions that output to two registers.

TIHan committed:
Added CalculateChecksumTest

TIHan committed:
Only allow GPR

TIHan commented:
This is ready for review @dotnet/jit-contrib @BruceForstall - just awaiting diff results

BruceForstall commented:
[Diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=125588&view=ms.vss-build-web.run-extensions-tab)

Interesting that there are diffs and ~0.01% TP hit on x64, but no diffs on x86 though it still has the TP hit.

TIHan commented:
> Interesting that there are diffs and ~0.01% TP hit on x64, but no diffs on x86 though it still has the TP hit.

That is expected. The TP hit will happen regardless. No diffs on x86 because, well, there are no 64-bit registers, so the notion of if the upper 32-bits are zero doesn't exist.

tannergooding commented:
Is it worth putting `#if defined(TARGET_AMD64)` around the relevant logic to avoid the TP regression on x86?

TIHan commented:
> Is it worth putting #if defined(TARGET_AMD64) around the relevant logic to avoid the TP regression on x86?

I don't think so, the TP regression comes from how we store and index into the last instructions.

BruceForstall commented on src/coreclr/jit/emit.cpp:
```diff
@@ -1005,19 +1005,20 @@ insGroup* emitter::emitSavIG(bool emitAdd)
 
-        emitLastIns = (instrDesc*)((BYTE*)id + ((BYTE*)emitLastIns - (BYTE*)emitCurIGfreeBase));
+        emitLastInstrs[emitInsCount % ArrLen(emitLastInstrs)] =
+            (instrDesc*)((BYTE*)id + ((BYTE*)emitGetLastIns() - (BYTE*)emitCurIGfreeBase));
```
There's a bug here:

We fill a static buffer with instrDesc's. When the buffer is full, or we move to a new BasicBlock with a label, we save that by allocating memory and moving them to new memory. This code updates the `emitLastIns` pointer to point to that new memory. Your new code only moves the last one, but doesn't update all the other instructions that have been moved. You need some kind of loop of size `emitCurIGinsCnt` to move them all.

TIHan replied:
This might make this PR more complicated. If I need to update all the previous instructions, I have a feeling that TP is going to suffer. We should discuss this offline.

kunalspathak replied:
Hhm, my impression was that whenever we see new IG, we would reset the `lastIns` table (to make sure TP is not impacted because of this, it was planned to just keep reset the current pointer and the no. of valid entries in the table and overwrite existing table), isn't that the case? I had related question around that - if my understanding is correct, how reliably can we check about `emitHasLastIns()` , etc.?

kunalspathak replied:
Also, you won't be checking cross IG last instructions, so why should we update this to point to new buffer?

TIHan replied:
I updated the PR to update the last instructions to point to the new memory, the TP diff is bad:

After: with updating
> [11:14:22] Total instructions executed delta: 144653690 (0.12% of base)

vs.

Before: without updating
> [11:17:27] Total instructions executed delta: 22510491 (0.02% of base)

--

I'm going to look into other ways to handle this. We really can't afford to update all instructions - instead, when we save the IG, we should just reset what is allowed to be enumerated in the last instrs array.

TIHan replied:
Trying something different, I added `emitCurLastInsCnt` to keep track of the count. When we update the last instr here, I reset the count to `1`.

BruceForstall commented on src/coreclr/jit/emit.h:
```diff
@@ -2173,22 +2173,71 @@ class emitter
+
+        // We use 'emitCurIGinsCnt' because we do not want to reference an instruction
+        // from a previous IG as to protect from GC holes.
+        for (unsigned i = 0; i < emitCurIGinsCnt; i++)
```
Because this is based on `emitCurIGinsCnt`, it doesn't allow peeking back from IGF_EXTEND groups to previous groups.

We should allow that, since overflow groups are somewhat arbitrarily sized.

One way to do that might be to create a new `emitCurIGinsCntExtended` (or some other name) variable that counts all the instructions in a non-IGF_EXTEND IG and all subsequent IGF_EXTEND IGs (thus, is zeroed out when we create a new non-IGF_EXTEND IG).

TIHan replied:
We should do this in a follow-up PR to avoid making this one more complicated; but I agree we should definitely do this.

TIHan replied:
Since I added `emitCurLastInsCnt` , it behaves like you said for `emitCurIGinsCntExtended`. However, actual asm diffs are small when the `emitLastInstrs` includes instructions from the current IG when creating a new extended IG. But, at least there won't be inconsistencies.

BruceForstall commented on src/coreclr/jit/emitxarch.cpp:
```diff
@@ -478,70 +478,151 @@ bool emitter::IsFlagsAlwaysModified(instrDesc* id)
-            {
-                return false;
-            }
+            case IF_RRD_CNS:
```
Why do we need to worry about the register read cases?

TIHan replied:
Is this only a register read case? I was assuming anything that is `IF_R*` is something that we write to.

BruceForstall replied:
We have `RRD`, `RWR`, `RRW` -- the first is a read, the second a write, and the third a read and write.

BruceForstall commented on src/coreclr/jit/emitxarch.cpp:
```diff
@@ -478,70 +478,151 @@ bool emitter::IsFlagsAlwaysModified(instrDesc* id)
+                    return PEEPHOLE_ABORT;
+                }
 
 #ifdef TARGET_AMD64
```
Shouldn't this entire function be under `ifdef TARGET_AMD64`? After all, x86 can't have upper 32 bits zero ever?

TIHan replied:
At the moment we don't need to, we only use `AreUpper32BitsZero` once:
```cpp
#ifdef TARGET_64BIT
        case GenIntCastDesc::ZERO_EXTEND_INT:
        case GenIntCastDesc::LOAD_ZERO_EXTEND_INT:
            ins     = INS_mov;
            insSize = 4;
            canSkip = compiler->opts.OptimizationEnabled() && emit->AreUpper32BitsZero(srcReg);
            break;
        case GenIntCastDesc::SIGN_EXTEND_INT:
        case GenIntCastDesc::LOAD_SIGN_EXTEND_INT:
            ins     = INS_movsxd;
            insSize = 4;
            break;
#endif
```

BruceForstall replied:
We don't need to, but wouldn't it document it better as never applying to x86? There's never a case it will apply to x86 in the future, either, right?

TIHan commented on src/coreclr/jit/emitxarch.cpp:
```diff
@@ -478,70 +478,151 @@ bool emitter::IsFlagsAlwaysModified(instrDesc* id)
 {
+    // Only allow GPRs.
+    // If not a valid register, then return false.
+    if (!((reg >= REG_INT_FIRST) && (reg <= REG_INT_LAST)))
```
Yea, that's a lot better :D

TIHan committed:
Update src/coreclr/jit/emitxarch.cpp

Co-authored-by: Bruce Forstall <brucefo@microsoft.com>

TIHan commented on src/coreclr/jit/emit.h:
```diff
@@ -2173,22 +2173,71 @@ class emitter
+
+    // Visits the last emitted instructions.
+    // Must be safe to do - use emitCanPeepholeLastIns for checking.
+    template <typename Action>
```
Fair, it's just a function: `*instrDesc -> emitPeepholeResult` . I would prefer it not be a template, but there isn't a way to type a parameter as a function(not function pointer) without using `std::function` as the type.

TIHan commented on src/coreclr/jit/emitxarch.cpp:
```diff
@@ -478,70 +478,151 @@ bool emitter::IsFlagsAlwaysModified(instrDesc* id)
-        case IF_RWR_MRD:
-        case IF_RWR_SRD:
-        case IF_RWR_ARD:
+    emitPeepholeLastInstrs([&](instrDesc* id) {
```
Hmm, maybe the function `emitPeepholeLastInstrs` isn't the best name. All this is doing is iterating over the previous instructions.

TIHan commented on src/coreclr/jit/emit.h:
```diff
@@ -2173,22 +2173,71 @@ class emitter
     }
 
-    instrDesc* emitLastIns;
+#define EMIT_MAX_IG_INS_COUNT 256
```
I don't think we will, it was just to make it symmetrical with the IG. When testing, I've tried values 16, 32, 64 and I think `32` was the sweet spot - there wasn't any improvements to codegen from 32 to 64.

kunalspathak replied:
Can we switch it to 32 then?

TIHan commented:
> If the x86 "last instruction buffer" was a single entry would it help? (Assuming there are no other x86 peeps that want the additional depth now)

I don't think it is worth the maintenance to have different implementations. Also, I will soon make a new PR that *might* improve TP across the board concerning `emitAllocAnyInstr` that would cover the TP cost here.

TIHan commented on src/coreclr/jit/emit.cpp:
```diff
@@ -1005,19 +1005,20 @@ insGroup* emitter::emitSavIG(bool emitAdd)
 #if defined(TARGET_XARCH)
-        assert(emitLastIns != nullptr);
-        if (emitLastIns->idIns() == INS_jmp)
+        assert(emitHasLastIns());
```
Ah, I see it too.

kunalspathak commented on src/coreclr/jit/emit.h:
```diff
@@ -2173,22 +2173,71 @@ class emitter
+
+    inline instrDesc* emitGetLastIns() const
+    {
+        return emitLastInstrs[emitInsCount % ArrLen(emitLastInstrs)];
```
```suggestion
        return emitLastInstrs[emitInsCount % EMIT_MAX_IG_INS_COUNT];
```

TIHan replied:
I prefer `ArrLen` over the `#define` - it will still emit a constant either way. The reason being is that you only have to specify the length of `emitLastInstrs` once:
```cpp
instrDesc* emitLastInstrs[EMIT_MAX_LAST_INSTR_COUNT];
```

TIHan committed:
Feedback

TIHan committed:
Feedback

TIHan committed:
Merge remote-tracking branch 'upstream/main' into mov-elim-1

TIHan committed:
Updating last instructions when saving IG. Using 'min(emitCurIGinsCnt, EMIT_MAX_LAST_INS_COUNT)' when iterating last instructions.

TIHan committed:
Added 'emitCurLastInsCnt'

TIHan committed:
Remove commented code

TIHan committed:
Using 'emitCurLastInsCnt' for the ring buffer

TIHan committed:
Formatting

TIHan committed:
Fix last instructions if the next IG is extended

TIHan commented:
@BruceForstall This is ready for review again, pending CI. 

I added fixing more than one last instruction if we are about to create a new extended IG. The throughput is not as bad, about 0.05% compared to 0.12% if we fixed up the last instructions all the time.

TIHan committed:
Update emit.cpp

TIHan committed:
Update emit.cpp

TIHan commented:
After some thought, I think being able to look back further than one instruction across extended IGs might be risky for GC holes; so I reverted the change to the original behavior when fixing up the last emitted instructions on IG save.

TIHan committed:
Merged

TIHan committed:
Fix build

TIHan committed:
Fix build

TIHan committed:
Fix build

TIHan committed:
Formatting

BruceForstall commented:
@TIHan There are assertions crossgen'ing System.Private.CoreLib.dll, e.g.:
```
Assertion failed 'emitHasLastIns() == (emitLastInsIG == nullptr)' in 'Microsoft.Win32.SafeHandles.CriticalHandleMinusOneIsInvalid:get_IsInvalid():bool:this' during 'Generate code' (IL size 15; hash 0x92df4c55; FullOpts)
```

BruceForstall commented on src/coreclr/jit/emitarm64.cpp:
```diff
@@ -16069,11 +16070,12 @@ bool emitter::IsRedundantLdStr(
-    insFormat lastInsfmt = emitLastIns->idInsFmt();
-    emitAttr  prevSize   = emitLastIns->idOpSize();
-    ssize_t prevImm = emitLastIns->idIsLargeCns() ? ((instrDescCns*)emitLastIns)->idcCnsVal : emitLastIns->idSmallCns();
+    regNumber prevReg1   = emitGetLastIns()->idReg1();
```
Maybe it would be better to add:
```
    const emitLastIns* = emitGetLastIns();
```
instead of calling the function many times?

BruceForstall replied:
A general pattern might be:
```
if (!emitHasLastIns())
{
    return;
}
const emitLastIns* = emitGetLastIns();
...
```
or, similarly,
```
if (emitHasLastIns())
{
    const emitLastIns* = emitGetLastIns();
    ...
}
```

TIHan replied:
I think the CPP compilers will do this already right? `emitGetLastIns` is small so it will be inlined and a CSE could be picked up.

TIHan replied:
True, I can fix that.

TIHan committed:
Fix assertion

BruceForstall commented on src/coreclr/jit/emit.cpp:
```diff
@@ -1021,27 +1021,28 @@ insGroup* emitter::emitSavIG(bool emitAdd)
 
-        emitLastIns   = (instrDesc*)((BYTE*)id + ((BYTE*)emitLastIns - (BYTE*)emitCurIGfreeBase));
-        emitLastInsIG = ig;
+        emitLastInstrs[0] = (instrDesc*)((BYTE*)id + ((BYTE*)emitGetLastIns() - (BYTE*)emitCurIGfreeBase));
+        emitCurLastInsCnt = 1;
```
I think there's a fundamental problem with capping the last instruction buffer when we hit an IG boundary: we will get Checked/Release asm diffs.

The problem is that the IG buffer of instrDescs is filled with different numbers of instructions in Checked and Release and so different optimizations will occur when that boundary is crossed.

I think you need to update all the saved instrs if `emitAdd` is true. If it is false, you can cap it to one, as done here.

BruceForstall replied:
We have to work across IGF_EXTEND IG boundaries for correctness. TP needs to be considered in that context.

TIHan replied:
Ok, I will add back the update loop.

BruceForstall commented:
Note: https://github.com/dotnet/runtime/pull/77540 is going to interact with this PR

TIHan commented:
Thank you for the second review @BruceForstall .

> Note: https://github.com/dotnet/runtime/pull/77540 is going to interact with this PR

That's good to know, will keep it in mind. The PR you made last week also touched this area as well, but I managed to resolve the conflicts.

BruceForstall commented on src/coreclr/jit/emit.h:
```diff
@@ -2175,19 +2175,34 @@ class emitter
     // The emitForceNewIG check here prevents peepholes from crossing nogc boundaries.
     // The final check prevents looking across an IG boundary unless we're in an extension IG.
-    bool emitCanPeepholeLastIns()
+    bool emitCanPeepholeLastIns() const
```
What does it mean to check this function for the "last" instruction, when there are _N_ last instructions to check?

Also, there is a check on `emitLastInsIG`. But the `N` last instructions might be spread over multiple IGs.

TIHan replied:
You think the check is not sufficient for checking back "N" instructions?

My line of thinking was, if I can peephole the last instruction, we should be able to check back further.

If we definitely cannot, then when we save the IG we need to reset the count if we know that it's not safe between IG boundaries, even if the IG is extended.

BruceForstall replied:
> You think the check is not sufficient for checking back "N" instructions?

Well, if the "previous instruction list" crosses multiple IG boundaries, then I think this check would need to be written to take that into account -- maybe by checking for each instruction we look back.

TIHan commented:
Putting this is draft mode for now as there are changes being made in the same area.

TIHan committed:
merging

TIHan committed:
Fixed minor issue

TIHan committed:
Merging complete

TIHan committed:
Fix build

TIHan committed:
Added canPeepholeCrossIGBoundaries

TIHan committed:
Cleanup

TIHan committed:
More cleanup

TIHan committed:
More cleanup

TIHan committed:
More cleanup

TIHan committed:
More cleanup

TIHan committed:
More cleanup

TIHan commented:
@BruceForstall - This is ready for review pending CI.

TIHan committed:
Fix build

TIHan committed:
Add count back

TIHan committed:
Formatting

BruceForstall commented:
@TIHan Looks like there are a bunch of arm64 asserts.

TIHan commented:
Yea, there might be an issue with backwards nav on arm64

TIHan committed:
Merge remote-tracking branch 'upstream/main' into mov-elim-1

TIHan committed:
Merge branch 'main' into mov-elim-1

TIHan committed:
Added back comments

TIHan committed:
Turn off backwards nav for arm

TIHan committed:
Added debug check for backward nav checks

TIHan committed:
Remove debug check

TIHan committed:
Merge remote-tracking branch 'upstream/main' into mov-elim-1

BruceForstall commented:
[Diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=172425&view=ms.vss-build-web.run-extensions-tab)

BruceForstall commented on src/coreclr/jit/emit.h:
```diff
@@ -2288,31 +2292,134 @@ class emitter
+        return (emitLastIns != nullptr);
+    }
+
+    // Checks to see if we can cross between the two given IG boundaries.
```
This function is very confusing. It would benefit from a full, normal function header.

In general, here and elsewhere in your change, you use the term `prevIG` which is confusing, because you use it to mean "the IG of the previous instruction", not "the previous IG in the IG list". I think you should rename it to be clear, e.g., `prevInsIG`. And, to be clear, here you could have `curInsIG` instead of just `ig`.

BruceForstall replied:
Another thing is, it is called even if there is no "IG boundary".

Maybe name it `isInsIGSafeForPeepholeOptimization`?

TIHan replied:
Sure, I will try to refactor some of this.

BruceForstall commented on src/coreclr/jit/emit.h:
```diff
@@ -2288,31 +2292,134 @@ class emitter
+    // We have the following checks:
+    // 1. Looking backwards across an IG boundary can only be done if we're in an extension IG.
+    // 2. The IG of the previous instruction must have the same GC interrupt status as the current IG.
+    //    This is related to #2; it disallows peephole when the previous IG is GC and the current is NOGC.
```
This `#2` doesn't make sense anymore since it was copy-paste during the refactoring.

TIHan replied:
Ok, will update it.

BruceForstall commented on src/coreclr/jit/emit.h:
```diff
@@ -2288,31 +2292,134 @@ class emitter
+    //    This is related to #2; it disallows peephole when the previous IG is GC and the current is NOGC.
+    inline bool canPeepholeCrossIGBoundaries(insGroup* prevIG, insGroup* ig) const
+    {
+        if (emitCurIG == ig)
+        {
+            return (((emitCurIGinsCnt > 0) || (ig->igFlags & IGF_EXTEND)) && // We're not at the start of a new
+                                                                             // IG or we are at the start of a
+                                                                             // new IG, and it's an extension IG.
+                    ((prevIG->igFlags & IGF_NOGCINTERRUPT) == (ig->igFlags & IGF_NOGCINTERRUPT)));
+        }
+        else if (prevIG != ig)
+        {
+            return ((prevIG->igFlags & IGF_EXTEND) &&
+                    ((prevIG->igFlags & IGF_NOGCINTERRUPT) == (ig->igFlags & IGF_NOGCINTERRUPT)));
+        }
+        return true;
```
Assuming you rename the arguments `prevInsIG` and `curInsIG` as suggested, I think this would be more clear:

```suggestion
        if (prevInsIG == curInsIG)
        {
             return true;
        }
        else 
        {
            return (curInsIG->igFlags & IGF_EXTEND) &&
                    ((prevInsIG->igFlags & IGF_NOGCINTERRUPT) == (curInsIG->igFlags & IGF_NOGCINTERRUPT));
        }
```

BruceForstall commented on src/coreclr/jit/emitxarch.cpp:
```diff
@@ -487,12 +487,6 @@ bool emitter::IsFlagsAlwaysModified(instrDesc* id)
-//
-//    movsx eax, ... might seem viable but we always encode this
-//    instruction with a 64 bit destination. See TakesRexWPrefix.
-
```
I think both `emitter::AreUpper32BitsZero()` and `emitter::AreUpper32BitsSignExtended` need to be under `#ifdef TARGET_64BIT`. They simply don't make sense otherwise.

TIHan replied:
I have another PR that does this: https://github.com/dotnet/runtime/pull/82276

TIHan commented on src/coreclr/jit/emit.h:
```diff
@@ -2288,31 +2292,134 @@ class emitter
+        {
+            assert(id != nullptr);
+
+#if !defined(TARGET_LOONGARCH64)
```
Ok, I can remove them.

TIHan commented on src/coreclr/jit/emit.h:
```diff
@@ -2288,31 +2292,134 @@ class emitter
+#if !defined(TARGET_LOONGARCH64)
+            switch ((ID_OPS)emitFmtToOps[id->idInsFmt()])
+            {
+                // We cannot safely look at previous instructions at these boundaries.
```
This is being pessimistic, but it's definitely *true* for determining if the upper 32 bits are zero.

Would you rather have this kind of check in `AreUpper32BitsZero`?

BruceForstall replied:
Can you tell me _why_ "we cannot safely look at previous instructions at these boundaries"? Even for "AreUpper32BitsZero"?

TIHan replied:
We can't determine if the register's upper 32-bits are zero when the previous instruction is a branching operation because we cannot determine the control flow of the register at this point, so we have to bail.

BruceForstall replied:
I guess I don't understand that. If you have:
```
movzx rax, eax (1)
<conditional branch> L_123 // branch somewhere
movzx rax, eax // (2)
```
then if control reaches (2), the upper 32 bits of rax are guaranteed to be zero. The branch doesn't affect them.

What does matter is if there is a branch *target* between (1) and (2). But that is handled by not allowing peeps to cross non-IGF_EXTEND IG boundaries.

TIHan replied:
It seems plausible, if we know there isn't a label in-between:
```
movzx rax, eax (1)
<conditional branch> L_123 // branch somewhere
<label>
movzx rax, eax // (2)
```
^ - I wouldn't deem this safe to skip.

TIHan committed:
Feedback. Allow looking back at conditional ops.

TIHan committed:
Update src/coreclr/jit/emit.h

Co-authored-by: Bruce Forstall <brucefo@microsoft.com>

TIHan commented on src/coreclr/jit/emit.h:
```diff
@@ -2288,31 +2292,111 @@ class emitter
+                    insGroup* prevIG = ig;
+                    if (emitPrevID(ig, id))
+                    {
+                        if (isInsIGSafeForPeepholeOptimization(prevIG, ig))
```
The original order is correct, but I can change the name.

TIHan committed:
Feedback

TIHan committed:
Feedback

BruceForstall commented on src/coreclr/jit/emit.h:
```diff
@@ -2288,31 +2292,111 @@ class emitter
+                    insGroup* savedInsIG = curInsIG;
+                    if (emitPrevID(curInsIG, id))
+                    {
+                        if (isInsIGSafeForPeepholeOptimization(savedInsIG, curInsIG))
```
Why do you say this argument ordering is correct?

The function signature is:
```
inline bool isInsIGSafeForPeepholeOptimization(insGroup* prevInsIG, insGroup* curInsIG) const
```

Thus, the first argument, `prevInsIG` is the IG for the previous (earlier in code sequence) instruction, and `curInsIG` is the IG for the current instruction IG.

Here, the `savedInsIG` is the "later in code sequence" instruction, since `emitPrevID` moved `curInsIG` backwards.

TIHan replied:
Ah I see it now, ok will fix the orderring.

TIHan committed:
Fix ordering

TIHan committed:
Added regression test. div,idiv,mul,imul should be noted as write to ops.

TIHan committed:
Added another regression test

TIHan committed:
Actually added the regression test

BruceForstall commented on src/coreclr/jit/emit.h:
```diff
@@ -2288,31 +2292,111 @@ class emitter
+    {
+        assert(emitHasLastIns() == (emitLastInsIG != nullptr));
+
+        return emitHasLastIns() && // there is an emitLastInstr
```
Maybe this is for a separate PR, but I think we need to prevent peephole optimizations when we're in the prolog or epilog. I.e.,
```
 if (emitIGisInProlog(emitCurIG) || emitIGisInEpilog(emitCurIG))
    {
        return false;
    }
#ifdef FEATURE_EH_FUNCLETS
    if (emitIGisInFuncletProlog(emitCurIG) || emitIGisInFuncletEpilog(emitCurIG))
    {
        return false;
    }
#endif
```
There is too much special handling in the prolog/epilog (e.g., unwinding) to allow peeps to kick in. There may be very specific cases where they are ok, but that requires some careful thinking.

TIHan committed:
Fix test

TIHan commented:
/azp run runtime-coreclr superpmi-replay

TIHan committed:
Merged

TIHan committed:
Fix merge

TIHan commented:
/azp run runtime-coreclr gcstress0x3-gcstress0xc

TIHan commented:
/azp run jitstress

TIHan commented:
/azp run runtime-coreclr jitstress

TIHan commented:
@dotnet/jit-contrib @BruceForstall This is ready again. I ran gcstress and jistress, they passed CI. The currently failures are unrelated.

AndyAyersMS commented:
Is there any way to avoid the TP impact in min opts?

BruceForstall commented:
[Diffs](https://dev.azure.com/dnceng-public/public/_build/results?buildId=182672&view=ms.vss-build-web.run-extensions-tab)

BruceForstall commented:
> Is there any way to avoid the TP impact in min opts?

I believe it's a result of enabling backwards navigation in the insGroup/instrDesc (https://github.com/dotnet/runtime/pull/80840), and maintaining those data structures. It's not obvious how this could only be done for non-MinOpts (and whether it would be advisable if it could be done).

BruceForstall reviewed (approved):
LGTM
